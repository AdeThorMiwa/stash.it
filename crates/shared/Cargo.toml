[package]
name = "shared"
version = "0.1.0"
edition = "2024"

[dependencies]
uuid = { version = "1.18.1", features = ["v4"] }
thiserror = { workspace = true }
serde = { workspace = true }
chrono = { workspace = true }
async-trait = { workspace = true }
alloy = "1.0.41"
more-di = { workspace = true }
tokio = { workspace = true }
config = { workspace = true }

[features]
testing = []
