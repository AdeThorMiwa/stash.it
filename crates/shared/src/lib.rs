pub mod domain;
pub mod infrastructure;
#[cfg(feature = "testing")]
pub mod testing;
