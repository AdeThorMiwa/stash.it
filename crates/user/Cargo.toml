[package]
name = "user"
version = "0.1.0"
edition = "2024"

[dependencies]
shared = { path = "../shared" }
macros = { path = "../macros" }
email_address = "0.2.9"
serde = { workspace = true }
chrono = { workspace = true }
async-trait = { workspace = true }
more-di = { workspace = true }
once_cell = { workspace = true }
send_wrapper = { workspace = true }
ctor = { workspace = true }
rand = "0.9.2"
thiserror = { workspace = true }
jsonwebtoken = { version = "10.0.0", features = ["aws_lc_rs"] }
derive_builder = { workspace = true }
config = { workspace = true }
tokio = { workspace = true }
regex = { workspace = true }
