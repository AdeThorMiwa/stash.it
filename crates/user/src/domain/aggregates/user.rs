use crate::domain::value_objects::{email::EmailAddress, user_status::UserStatus};
use chrono::Utc;
use shared::domain::value_objects::{date::Date, pid::Pid};

#[derive(Debug, Clone)]
pub struct User {
    pub(crate) pid: Pid,
    pub(crate) email: EmailAddress,
    pub(crate) status: UserStatus,
    pub(crate) created_at: Date,
    pub(crate) last_login_at: Date,
}

impl User {
    /// create a new user
    pub fn new(email: EmailAddress) -> Self {
        let pid = Pid::new();

        Self {
            pid,
            email,
            status: UserStatus::PendingProfile,
            created_at: Utc::now(),
            last_login_at: Utc::now(),
        }
    }

    /// update user status to `new_status`
    pub fn update_status(&mut self, new_status: UserStatus) {
        self.status = new_status;
    }

    /// update user last login time to now
    pub fn update_last_login(&mut self) {
        self.last_login_at = Utc::now()
    }
}
