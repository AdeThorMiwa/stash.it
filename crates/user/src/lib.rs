use macros::initialize_container;

pub mod application;
mod domain;
mod infrastructure;

initialize_container!();

