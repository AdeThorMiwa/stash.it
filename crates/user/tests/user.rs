use crate::common::{bootstrap::bootstrap, prepare::prepare_authenticated_user};
use insta::{assert_debug_snapshot, with_settings};
use shared::{
    configure_insta, domain::value_objects::wallet_address::WalletAddress, infrastructure::types::Result,
    testing::insta_filters::redactions::cleanup_model_generics,
};
use std::{str::FromStr, sync::Arc};
use user::{
    application::user::{UserManagementService, command::CreateUserProfileCommand},
    domain::value_objects::display_name::DisplayName,
};

mod common;

#[tokio::test]
async fn can_create_user_profile() -> Result<()> {
    // Arrange
    configure_insta!();
    let provider = bootstrap();
    let user_service: Arc<UserManagementService> = provider.get_required();
    let (pid, ..) = prepare_authenticated_user(&provider).await?;
    let create_profile_command = CreateUserProfileCommand {
        user_id: pid.clone(),
        display_name: DisplayName::from_str("testuser").unwrap(),
        wallet_address: WalletAddress::from_str("0x52471a768b76B8cC647f2F28198cB0E44C38C2cF").unwrap(),
    };

    // Act
    let profile = user_service.create_user_profile(create_profile_command).await?;

    // Assert
    assert_eq!(profile.get_user_id(), &pid, "User id on profile must match user pid");
    with_settings!({
        filters => cleanup_model_generics(),
        snapshot_suffix => "user_profile"
    }, {
        assert_debug_snapshot!(profile);
    });

    Ok(())
}
