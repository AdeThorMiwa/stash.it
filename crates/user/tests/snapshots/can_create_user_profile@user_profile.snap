---
source: crates/user/tests/user.rs
expression: profile
---
Profile {
    pid: Pid(
        PID,
    ),
    user_id: Pid(
        PID,
    ),
    display_name: DisplayName(
        "testuser",
    ),
    wallet_address: WalletAddress(
        "0x52471a768b76B8cC647f2F28198cB0E44C38C2cF",
    ),
}
